### PR TITLE
issue-editor2-2700-options-to-move-gui-node-up-down-should-be-in-context-menu

### PR DESCRIPTION
**Pull request by madwareru**
Closes defold/editor2-issues#2700

Added context menu so gui nodes are possible to move up/down with a context menu and user faster figure out hotkey for it. Optional side effect: seems to affect atlas outline too so we can flip sprites in animation with a context menu too

### CODE REVIEW

**Comment by madwareru** [comment]
@matgis-king please review this (somehow I'm not allowed to request your review other way)

**Review by vlaaad — CHANGES_REQUESTED** [review]
Hi @madwareru, thanks for your contribution!

I have a couple of concerns about these context menu items:
- I think they should not be at the very top of a context menu. I would expect more commonly used options to be there, while Move Up/Down commands are there mostly to teach a user that this ability exists. My gut feeling tells me it should be below the cut/copy/paste section. Or above. What do you guys think, @matgis-king @eangelin @bagucode?
- It would be nice if these options could be shown as disabled (aka visible, but greyed out) when the move is not possible — for example when the item is first or last in the list. This can be achieved by adding `(enabled? ...)` clause to command definition.

**Comment by madwareru** [comment]
Thanks Vlad, I agree, position for these menu items requires some discussion, so I will wait what other reviewers say and then try to position it in appropriate place.

> It would be nice if these options could be shown as disabled (aka visible, but greyed out) when the move is not possible — for example when the item is first or last in the list. This can be achieved by adding `(enabled? ...)` clause to command definition.

Interesting. Seems like a handy feature.

**Comment by madwareru** [comment]
I just updated PR to support `(enabled? ...)` feature.
Still waiting what other reviewers think about positioning of context menu items.

File: editor/src/clj/editor/atlas.clj [hunk 1]
```diff
@@ -777,8 +777,18 @@
 
 (handler/defhandler :move-up :workbench
   (active? [selection] (move-active? selection))
+  (enabled? [selection] (let [node-id (selection->image selection)
+                              parent (core/scope node-id)
+                              children (vec (g/node-value parent :nodes))
+                              node-child-index (.indexOf ^java.util.List children node-id)]
+                          (> node-child-index 0)))
```
Comment (vlaaad) [inline editor/src/clj/editor/atlas.clj:784]: Please use `(pos? x)` instead of `(> x 0)`

File: editor/src/clj/editor/atlas.clj [hunk 2]
```diff
@@ -777,8 +777,18 @@
...
   (run [selection] (move-node! (selection->image selection) -1)))
 
 (handler/defhandler :move-down :workbench
   (active? [selection] (move-active? selection))
+  (enabled? [selection] (let [node-id (selection->image selection)
+                              parent (core/scope node-id)
+                              children (vec (g/node-value parent :nodes))
+                              node-child-index (.indexOf ^java.util.List children node-id)]
+                          (< node-child-index (- (.size children) 1))))
```
Comment (vlaaad) [inline editor/src/clj/editor/atlas.clj:793]: Please use `(dec x)` instead of `(- x 1)`

**Comment by madwareru** [comment]
Thanks @vlaaad! Fixed :) But there are some strange trouble in CI. I'll try to make a fresh build and test it on my machine today

**Comment by britzl** [comment]
I've restarted the CI job. Let's see if it is only a temporary hiccup (it has happened before with GitHub CI).

**Review by vlaaad — APPROVED** [review]
Tested it locally — works wonderfully!

Please update remaining uses of `(> x 0)` to `(pos? x)` and `(- x 1)` to `(dec x)`

**Pull request merged**